### PR TITLE
[WIP] Reduce ThreadLocal<T>.LinkedSlot creation by 64 fold

### PR DESCRIPTION
**Pull request by cshung**
This PR is meant to reduce allocation by limiting the number of LinkedSlot object instances.

Before the change, per instance of ThreadLocal, per thread accessed it, there will be an instance of a  LinkedSlot object that holds the slot.

For scenarios like https://github.com/dotnet/runtime/issues/2382, there will be a lot of LinkedSlot objects, and increased GC latency significantly.

After the change, we reduced that by 64 fold. The idea is to use a single object to host 64 slots, and we used some simple arithmetic to determine which spot to use and used a simple bitmask to determine if the spot is used so that we can reclaim the whole LinkedSlot when it is no longer used by anyone.

It took me a good while to figure out the bitmask operations are non-atomic. After fixing that, I am hoping this is good to go. Unfortunately, I cannot spend more time on this. Just in case if it doesn't work out, it would be great if this can be picked up and completed.

### CODE REVIEW

**Comment by Maoni0** [comment]
looping in @stephentoub

**Comment by stephentoub** [comment]
> where can I find the test

Tests for ThreadLocal are at https://github.com/dotnet/runtime/blob/master/src/libraries/System.Threading/tests/ThreadLocalTests.cs

> perf test

I don't think we have any yet (I'm sure they existed at one point, but not for .NET Core).  We've been adding microbenchmarks as we change stuff.  You'd probably want to add them in https://github.com/dotnet/performance/tree/master/src/benchmarks/micro/libraries/System.Threading.

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs [hunk 1]
```diff
@@ -187,26 +191,41 @@ protected virtual void Dispose(bool disposing)
...
+                        for (int i = 0; i < valuePerSlot; i++)
+                        {
+                            slotArray[slotId].Value!._value![i] = default!;
+                        }
+                        slotArray[slotId].Value = null;
+                    }
                 }
             }
             _linkedSlot = null;
+            // Experiment show that if I never return the id, the test case failure will not repro.
```
Comment (cshung) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]: This comment is accidental, I will remove it.

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs [hunk 2]
```diff
@@ -187,26 +191,41 @@ protected virtual void Dispose(bool disposing)
...
+                long created = 0;
                 for (LinkedSlot? linkedSlot = _linkedSlot._next; linkedSlot != null; linkedSlot = linkedSlot._next)
                 {
-                    LinkedSlotVolatile[]? slotArray = linkedSlot._slotArray;
+                    linkedSlot.Unset(withinSlotId);
+                    created |= linkedSlot._created;
+                }
 
-                    if (slotArray == null)
+                // Experiment show that if I return the id, whether or not I free the array, the test case still repros.
```
Comment (cshung) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]: Same here

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs [hunk 3]
```diff
@@ -175,6 +177,8 @@ protected virtual void Dispose(bool disposing)
             lock (s_idManager)
             {
                 id = ~_idComplement;
+                int slotId = id / valuePerSlot;
+                int withinSlotId = id % valuePerSlot;
```
Comment (khellang) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]: `Math.DivRem`? 🤔

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]: Note that `valuePerSlot` is a power-of-two constant. Do all JIT optimizations for power-of-two constants still kick in when this DivRem call?

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs [hunk 4]
```diff
@@ -27,6 +27,8 @@ namespace System.Threading
     [DebuggerDisplay("IsValueCreated={IsValueCreated}, Value={ValueForDebugDisplay}, Count={ValuesCountForDebugDisplay}")]
     public class ThreadLocal<T> : IDisposable
     {
+        private const int valuePerSlot = 64;
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]:
```suggestion
        private const int ValuePerSlot = 64;
```
https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md

**Comment by cshung** [comment]
The metric I am trying to improve is the GC latency in this given scenario (copied from https://github.com/dotnet/runtime/issues/2382):

```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace RavenTicket
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine($"GC ServerMode: {GCSettings.IsServerGC }, LOH Compact: {GCSettings.LargeObjectHeapCompactionMode}, Latency: {GCSettings.LatencyMode} ");

            int running = 0, stopped = 0;
            var abc = new string('*', 1024*2024);

            var task = Task.Run(() =>
            {
                var foregroundColor = Console.ForegroundColor;
                while (true)
                {
                    var sp = Stopwatch.StartNew();
                    GC.Collect(2);
                    var gc = sp.ElapsedMilliseconds;
                    sp.Restart();
                    GC.WaitForPendingFinalizers();
                    var run = Volatile.Read(ref running);
                    var stop = Volatile.Read(ref stopped);
                    if (gc > 200)
                    {
                        Console.ForegroundColor= ConsoleColor.Red;
                    }
                    Console.WriteLine($"{gc} - {sp.ElapsedMilliseconds} - {run} - {stop} - {run - stop:#,#;;0}");
                    if (gc > 200)
                    {
                        Console.ForegroundColor = foregroundColor;
                    }
                    Thread.Sleep(500);
                }
            });

            var list = new List<ThreadLocal<WeakReference>>();
            var threads = new List<Thread>();

            const int numberOfThreadLocals = 10_000;
            const int numberOfThreads = 2500;
            const int numberOfActiveThreads = 64;


            for (int i = 0; i < numberOfThreadLocals; i++)
            {
                list.Add(new ThreadLocal<WeakReference>());
            }
            var stageOne= new CountdownEvent(numberOfThreads);
            var stageTwo = new SemaphoreSlim(0, numberOfActiveThreads);
            for (int i = 0; i < numberOfThreads; i++)
            {
                var copy = i;
                threads.Add(new Thread(() =>
                {
                    for (var index = copy; index < list.Count; index += (copy % 10) + 1)
                    {
                        list[index].Value = null;
                    }
                    stageOne.Signal();
                    stageTwo.Wait();
                    Interlocked.Increment(ref running);
                    string s = null;
                    if (copy % 128 == 0)
                        s = AllocateLotsOfMemory();
                    for (var index = copy; index < numberOfThreadLocals; index += 16)
                    {
                        var t = list[index];
                        if (index % 2 == 0)
                        {
                            t.Value = new WeakReference(list);
                        }
                        else if ((index % 5) == 0)
                        {
                            t.Value = new WeakReference(abc);
                        }
                        else if ((index % 51) == 0)
                        {
                            t.Value =
                                new WeakReference(s);
                        }
                        else
                        {
                            t.Value =
                              new WeakReference(AllocateLotsOfMemory());
                        }
                    }

                    s = null;
                    stageTwo.Release();
                    Interlocked.Increment(ref stopped);
                    threads[copy].Join();
                })
                {
                    IsBackground = true
                });
            }
            threads.ForEach(t => t.Start());
            Console.WriteLine("Init...");
            stageOne.Wait();
            Console.WriteLine("Starting..");
            stageTwo.Release(64);
            


            Task.Run(() => // force some additional memory traffic
            {
                for (int i = 0; i < 200_000; i++)
                {
                    AllocateLotsOfMemory();
                }
            });

            task.Wait();
            Console.WriteLine(abc);
        }

        private static string AllocateLotsOfMemory()
        {
            var sb = new StringBuilder();
            for (int j = 0; j < 10_000; j++)
            {
                sb.Append(j);
            }

            var s1 = sb.ToString();
            return s1;
        }
    }
}
```

As far as I understand, this is trying to emulate what is happening in production. I am focusing on the first column, the number of milliseconds spent during `GC.Collect()`, so lower is better.

My analysis of the scenario reveals that the latency comes from the huge amount of objects. Therefore I came up with the idea for this PR.

Here is the result of running the scenario with and without this change.

**Before:**
```
GC ServerMode: False, LOH Compact: Default, Latency: Interactive
0 - 0 - 0 - 0 - 0
Init...
2484 - 0 - 0 - 0 - 0
896 - 0 - 0 - 0 - 0
668 - 0 - 0 - 0 - 0
676 - 0 - 0 - 0 - 0
840 - 0 - 0 - 0 - 0
Starting..
831 - 0 - 83 - 23 - 60
846 - 237 - 107 - 44 - 63
842 - 1944 - 113 - 49 - 64
844 - 0 - 113 - 49 - 64
852 - 199 - 113 - 49 - 64
844 - 0 - 113 - 49 - 64
851 - 110 - 113 - 49 - 64
846 - 34 - 113 - 49 - 64
850 - 2235 - 113 - 49 - 64
867 - 841 - 113 - 49 - 64
846 - 0 - 115 - 51 - 64
845 - 1593 - 116 - 53 - 63
850 - 0 - 123 - 61 - 62
850 - 3385 - 126 - 63 - 63
1236 - 0 - 164 - 101 - 63
1198 - 39 - 165 - 102 - 63
1189 - 694 - 175 - 112 - 63
851 - 536 - 179 - 115 - 64
853 - 84 - 191 - 127 - 64
855 - 31 - 198 - 137 - 61
850 - 746 - 214 - 151 - 63
853 - 979 - 221 - 158 - 63
...
```
**After:**
```
GC ServerMode: False, LOH Compact: Default, Latency: Interactive
0 - 0 - 0 - 0 - 0
160 - 0 - 0 - 0 - 0
Init...
Starting..
178 - 0 - 64 - 1 - 63
175 - 0 - 82 - 23 - 59
153 - 21 - 105 - 48 - 57
270 - 259 - 128 - 64 - 64
124 - 0 - 128 - 64 - 64
120 - 3565 - 128 - 64 - 64
125 - 21 - 128 - 64 - 64
180 - 0 - 128 - 64 - 64
238 - 0 - 128 - 64 - 64
119 - 0 - 128 - 64 - 64
126 - 0 - 128 - 64 - 64
141 - 444 - 129 - 65 - 64
141 - 652 - 130 - 66 - 64
141 - 0 - 130 - 67 - 63
232 - 1988 - 131 - 67 - 64
124 - 1013 - 138 - 74 - 64
142 - 339 - 145 - 83 - 62
123 - 702 - 151 - 88 - 63
144 - 334 - 156 - 94 - 62
129 - 1251 - 183 - 123 - 60
146 - 65 - 202 - 139 - 63
128 - 172 - 218 - 154 - 64
125 - 0 - 218 - 155 - 63
129 - 0 - 224 - 160 - 64
126 - 641 - 226 - 162 - 64
256 - 2298 - 234 - 171 - 63
...
```

The improvement in this particular scenario is obvious.

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs [hunk 5]
```diff
@@ -187,23 +191,36 @@ protected virtual void Dispose(bool disposing)
...
-                    slotArray[id].Value!._value = default;
-                    slotArray[id].Value = null;
+                        // Remove the reference from the LinkedSlot to the slot table.
+                        linkedSlot._slotArray = null;
+
+                        // And clear the references from the slot table to the linked slot and the value so that
+                        // both can get garbage collected.
+                        for (int i = 0; i < ValuePerSlot; i++)
+                        {
+                            slotArray[slotId].Value!._value![i] = default!;
```
Comment (VSadov) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]: >slotArray[slotId].Value!._value![i] = default! [](start = 28, length = 46)

Is it necessary to zero out individual values?

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs [hunk 6]
```diff
@@ -673,7 +715,36 @@ internal LinkedSlot(LinkedSlotVolatile[]? slotArray)
...
+
+            // A bit mask to represent is the slot created
+            internal long _created;
+
+            internal void Set(int withinSlotId)
+            {
+                while (true)
+                {
+                    long current_created = _created;
+                    long new_created = current_created | (1L << withinSlotId);
```
Comment (lpereira) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadLocal.cs]: Micro-opt nit: will the JIT hoist `1<<withinSlotId` out of the loop?

**Pull request closed without merging**